### PR TITLE
New cmder.exe args and shared install capability

### PR DESCRIPTION
**Pull request by daxgames**
## [1.3.6-pre1](https://github.com/cmderdev/cmder) (2018-03-01)

**Fixed bugs:**

* Fixed Git version check recently added to master.

**Updates:**

* Modified Cmder tasks in default Conemu.xml to allow easily adding command line args for init.bat by adding some quotes.  This resulted in a ton of misc changes to this file.  See Adds below.
* Reworked `cmder.exe` command line argument handling to make it more flexible and easily added to.
* Reworked README.md tables to make them more readable in editors

**Implemented enhancements:**

* Append `%GIT_INSTALL_ROOT%\mingw32|64` folder to the path if it exists. Thanks: @gucong3000 
* Added `cmder.exe` command line args documenttion to `README.md`
* Added `:enhance_path` method to vendor\init.bat that modifies the path only if required.
   * To prepend: `call :enhance_path "%cmder_root%"`
   * to append: `call :enhance_path "%cmder_root%" append`
* Added `:enhance_path_recursive` method to vendor\init.bat that adds a path and all its sub directories to the path if required.  
   * Max recurse depth default is '1' configurable using `init.bat /max_depth [1-5]`. 6+ results in error. 
   * To prepend and go 3 levels deep: `call :enhance_path "%cmder_root%" 3`
   * To append and go 2 levels deep: `call :enhance_path "%cmder_root%" 2 append`
* Added ability to init.bat to accept command line args and documented them in README.md.  Allows users to change the behaviour of init.bat without editing the file.

| Argument                      | Description                                                                                      | Default                               |
| ----------------------------- | ----------------------------------------------------------------------------------------------   | ------------------------------------- |
| /c [user cmder root]          | Enables user bin and config folders for 'Cmder as admin' sessions due to non-shared environment. | not set                               |
| /d                            | Enables debug output.                                                                            | not set                               |
| /git_install_root [file path] | User specified Git installation root path.                                                       | '%CMDER_ROOT%\vendor\Git-for-Windows' |
| /home [home folder]           | User specified folder path to set `%HOME%` environment variable.                                 | '%userprofile%'                       |
| /max_depth [1-5]              | Define max recurse depth when adding to the path for `%cmder_root%\bin` and `%cmder_user_bin%`   | 1                                     |
| /svn_ssh [path to ssh.exe]    | Define %SVN_SSH% so we can use git svn with ssh svn repositories.                                | '%GIT_INSTALL_ROOT%\bin\ssh.exe'      |
| /user_aliases [file path]     | File path pointing to user aliases.                                                              | '%CMDER_ROOT%\config\user-liases.cmd' |
| /v                            | Enables verbose output.                                                                          | not set                               |

* Added new `cmder.exe /C \<path\>` argument
   * To use run Cmder.exe with "/C" command line argument. Example: `cmder.exe /C %userprofile%\cmder_config`
   * To use run with `Cmder as Admin` sessions you must specify "/c" command line argument to `init.bat` in tasks. See [README.md](./Readme.md) for details.
   * Enables shared Cmder install with Non-Portable Individual User Config
   * Supported by all supported shells (cmder, powershell, git bash, and external bash)
   * This will create the following directory structure if it is missing.

      ```
      c:\users\[username]\cmder_config
      ├───bin
      └───config
          └───profile.d
      ```

   * Shell init scripts run in the following order
      1. %cmder_root%\config\profile.d\*.[cmd|ps1|sh]
      1. %cmder_root%\config\user-profile.[cmd|ps1|sh]
      1. %userprofile%\cmder_config\config\profile.d\*.[cmd|ps1|sh]
      1. %userprofile%\cmder_config\config\user-profile.[cmd|ps1|sh]

### CODE REVIEW

**Comment by DRSDavidSoft** [comment]
@daxgames Great PR! One question though, does this remove the block where returned git version string is matched to verify its format?

**Comment by daxgames** [comment]
@DRSDavidSoft no

**Comment by daxgames** [comment]
@DRSDavidSoft Sorry for the initial short reply phone keyboard issues.

No,  I made the 'for' block that gets the version more intelligent so it:

1. Could deal with paths with spaces.  My cmder folder is in my 'Google Drive' folder so your code did not work because of the space.
2. Extracts the semantic version from the return value of `git --version` and passes it to the verify function.

**Comment by daxgames** [comment]
@DRSDavidSoft I do see something I need to fix though then I need you to test it in your system

**Comment by daxgames** [comment]
@DRSDavidSoft Can you please test and make sure it works for you?

**Comment by DRSDavidSoft** [comment]
@daxgames Checking now, will update this comment

File: vendor/init.bat [hunk 1]
```diff
@@ -6,8 +6,49 @@
 :: !!! THIS FILE IS OVERWRITTEN WHEN CMDER IS UPDATED
 :: !!! Use "%CMDER_ROOT%\config\user-profile.cmd" to add your own startup commands
 
-:: Set to > 0 for verbose output to aid in debugging.
-if not defined verbose-output ( set verbose-output=0 )
+:: Use -v command line arg or set to > 0 for verbose output to aid in debugging.
```
Comment (DRSDavidSoft) [inline vendor/init.bat]: Either change `-v` to `/v`, or make `init.bat` accept `-v` and `--git-install-root` formatted arguments as well as `/v` and `/git_install_root` format.

Personally I vote for the latter.

File: vendor/init.bat [hunk 2]
```diff
@@ -220,70 +292,71 @@ exit /b
...
-
-        :: now parse the version information into the corresponding variables
-        call :parse_version %~1 %~4
-
-        :: ... and maybe display it, for debugging purposes.
-        call :verbose-output Found Git Version for %~1: !%~1_MAJOR!.!%~1_MINOR!.!%~1_PATCH!.!%~1_BUILD!
-
-    ) else (
-        :: invalid format returned, use the vendored git instead
-        echo Invalid git version at "%git_executable%" detected!
```
Comment (DRSDavidSoft) [inline vendor/init.bat]: With the removal of this section, invalid formatted return string is not caught.

**Valid case:** if user has _any other_ executable in path accidentally named `git.exe` that won't return `git version xxxxx` on issuing `git.exe --version` on them.

IMO, `init.bat` should also verify if the return string has `git version` in it.

File: vendor/init.bat [hunk 3]
```diff
@@ -220,70 +292,71 @@ exit /b
...
     if not exist "%git_executable%" (
-        :: return a negative error code if the executable doesn't exist
+        call :verbose-output "%git_executable%" does not exist!
         exit /b -255
     )
 
     :: get the git version in the provided directory
-    for /F "delims=" %%F in ('%git_executable% --version 2^>nul') do @(
-        set "GIT_VERSION_%~1=%%F"
+    for /F "tokens=1,2,3 usebackq" %%F in (`"%git_executable%" --version 2^>nul`) do @(
```
Comment (DRSDavidSoft) [inline vendor/init.bat]: Yes, my mistake. Should have used `usebackq` with that syntax to avoid paths with space in their name. Damn you, MS Windows!

**Comment by DRSDavidSoft** [comment]
@daxgames I [reviewed your edits](https://github.com/cmderdev/cmder/pull/1665/files#diff-cabfc4f102e08adb66d90c726994933f) and tested it on my machine.

As you modified, the version comparing seems to also work perfectly on paths with spaces.

Aside from them, I also ran into an odd problem when executing `init.bat /v`:
Should this happen?

cmder_init_verboose

**Comment by daxgames** [comment]
@DRSDavidSoft. Yes, /v turns on verbose output.  Those are all calls to :verbose-output.  It certainly could be prettier.

Mainly a debugging thing to help us help people using under.

**Comment by daxgames** [comment]
Or maybe  after looking closer at other output.  Is cmder.exe you launched in c:\?

**Comment by DRSDavidSoft** [comment]
@daxgames no, I launched everything from `cmder/vendor` directory. As I launched `init.bat` directly, it seems that `cmder_root` was not set. Which makes me think:

1. It should detect if the variable is not set, instead of setting it to the root of the drive
2. It should set the variable accordingly.

This only happened when I turned on the verbose mode with the `/v` switch, and not when I manually set the variable `set verbose-output=1` in init.bat. that's what I was referring to, as odd :)

**Comment by daxgames** [comment]
Thats odd I don't see that but I have never runt it like you did.  I need to look at that.  What if you modify the task and add /v.  Look at the readme.md you will need to add additional set of "" around init.bat to make it work.

**Comment by DRSDavidSoft** [comment]
@daxgames Personally, I think [this](https://github.com/daxgames/cmder/pull/2) fixes the problem.

**Comment by daxgames** [comment]
I'll look at it.  Not where I can look t the code right now..

File: vendor/init.bat (hunk 1 above)
Comment (daxgames) [reply vendor/init.bat]: It is dos not linux/unix sticking with / cmder.exe uses /.  in batch to support - and -- and / is a lot of duplicated code.  fixed the mention of -v to be /v

**Comment by daxgames** [comment]
@DRSDavidSoft - I cannot duplicate the issue I saw with you manually running `init.bat /v`.

@cmderdev/trusted-contributors - Please test and comment, this has some pretty cool stuff in it.

File: vendor/init.bat (hunk 2 above)
Comment (daxgames) [reply vendor/init.bat]: fixed but slightly diff but it does what you want

Comment (DRSDavidSoft) [reply vendor/init.bat]: What did you change for this case?

**Comment by Stanzilla** [comment]
Looks good to me, what are the conflicts?

**Comment by daxgames** [comment]
@Stanzilla Conflicts?

**Comment by Stanzilla** [comment]
"This branch cannot be rebased due to conflicts"

**Comment by daxgames** [comment]
@Stanzilla I haven't seen that.  I just did a clean merge locally into a copy of master

**Comment by Stanzilla** [comment]
Hrrm, weird squash and merge and merge commit work, just rebase does not want to. Aw well, I think for PRs like that a merge commit is good anyway.

**Comment by DRSDavidSoft** [comment]
BTW @daxgames for now, let's keep `:enhance_path` and `:read_version` in the init.bat under the subroutines section. We'll split them into the `/bin` in future PRs.

**Comment by daxgames** [comment]
Hah too late... Been working on it already.  Actually not separating them out into bin rather added. `%cmder_root%\lib\lib_git.cmd|lib_path.cmd|lib_profile.cmd|lib_console.cmd` they can be 'included in any bat|cmd file by executing `call "%cmder_root%\lib\lib_git.cmd|lib_path.cmd|lib_profile.cmd|lib_console.cmd"`

And can be used by adding `%lib_console% debug-output init.bat "This is a debug message"`

Which produces output like:
```
DEBUG(init.bat): This is a debug message.

```

**Comment by DRSDavidSoft** [comment]
@daxgames Nice work then, actually! What branch are you committing these on?
Also, could you allow me to commit the `lib_git.cmd` so it keeps my handle in the history? Thanks! :)

**Comment by daxgames** [comment]
File ‘lib_git.cmd’ is attached as txt.

 

See :read_version and the integrated documentation(copied from enhance path – not complete) I started please fill it out for all subs.

 

This allows getting command line help for all subs by typing ‘lib_git.cmd /h’

 

Dax

**Comment by daxgames** [comment]
I haven’t committed anything yet.

**Comment by daxgames** [comment]
@DRSDavidSoft Sent you email with the lib_git.cmd attached so you can contribute it.  Needs to be clmpleted but it work.  see my email.

**Comment by daxgames** [comment]
See the 'lib-it' branch.

**Comment by daxgames** [comment]
Hah - the unlimited depth just bit me.  gotta fix that.

**Comment by daxgames** [comment]
@DRSDavidSoft @stanzilla - Max recurse depth added please test.

I also have this all broken into importable libraries in my 'lib-it' branch, mostly works just working on docs.

**Comment by daxgames** [comment]
@cmderdev/trusted-contributors I added a CHANGELOG.md entry to this PR and I called it `1.3.6-pre` and it links to `https://github.com/cmderdev/cmder` and not a release.  My suggestion is we just continue to add to this changelog entry until ready to 'release' and remove the `-pre` and link to the release when the release is built.  Hopefully that work for every one.

I am not sure what specific issue numbers this resolves or adds so if anyone doe pleas update with a PR to my branch at least until it is merged to master.

**Comment by daxgames** [comment]
Need to fix cmder /REGISTER|/UNREGISTER

**Comment by DRSDavidSoft** [comment]
@cmderdev/trusted-contributors Will this be merged soon?

**Comment by daxgames** [comment]
I need to fix the above first because it's broken because of the command line rework I did.

**Comment by Jackbennett** [comment]
Would you mind rebase and squashing as is reasonable for 1 contained item/fix/improvement per commit when you're ready dax.

**Comment by daxgames** [comment]
I will rebase and squash as much as possible but most of what is in here is related which is why its in here.

**Comment by Stanzilla** [comment]
I am on vacation for a few days so I won't do any testing soon but this looks really nice, great job!

**Comment by daxgames** [comment]
@DRSDavidSoft, @cmderdev/trusted-contributors - Cmder.exe arg processing fixed and squashed.  I think it's ready but please test.

**Comment by daxgames** [comment]
@cmderdev/trusted-contributors Still testing

**Comment by i-s-o** [comment]
@daxgames This is probably related to your last comment: I defined this ``GIT_INSTALL_ROOT=C:\tools\cmder\vendor\git-for-windows`` on my system, but cmder is still using the git installed by VS Community 2017. I worked around this by removing the VS2017-installed git from the ``%PATH%``, but VS2017 updated git recently, and git is back in the ``%PATH%`` again, breaking cmder vendored git.

**Comment by daxgames** [comment]
Fixed and ready to merge AFAIK - the issue mentioned above was releated to over zealous use of `setlocal enabledelayedexpansion` inside a `for` block that did not have corresponding `endlocal` calls and also due to the loop being broken with a `goto` thus even if there were enough `endlocal` calls it would have still failed.  Simply removed the offending extra `setlocal enabledelayedexpansion` to fix the issue.  @cmderdev/trusted-contributors and @DRSDavidSoft - PLEASE TEST!

**Comment by Stanzilla** [comment]
Can you rebase the branch, please? Otherwise, I am still on vacation and will start testing on the 12th.

**Comment by DRSDavidSoft** [comment]
@daxgames - My test results:
- Launching Cmder on an existing upgraded environment - _PASS!_
- Launching Cmder in a new environment - _PASS!_
- Calling `init.bat` from other Terminals than ConEmu - _PASS!_
- Git version comparing - _PASS!_
- Recursive Directory Path Enhancing - _PASS!_
- Path Enhancing Depth Limit - _PASS!_

Looks good to me! (P.S. I think `/home` argument should set _both_ `%HOME%` *and* `%USERPROFILE%`, otherwise I don't see the indication.)

@Stanzilla Please perform further (extensive) tests when you're back. Either way, this seems fine to merge, considering that this is a _pre1_ release. If releasing it uncovers any issues, we could fix them in the _rc_ release.

**Comment by daxgames** [comment]
@Stanzilla I will go ahead and merge it.  I assume that is what you meant by rebase?

**Comment by i-s-o** [comment]
I'm not sure if this is the right place to post this, I have a few issues:

1. After unzipping the latest cmder.7z from AppVeyor, running cmder.exe gives a bunch of error messages like:
``The system cannot find the batch label specified - enhance_path_recursive``
``The system cannot find the batch label specified - enhance_path``
AFAICT, this is because the lines in init.bat are terminated by LF instead of CRLF.

2. If I have ``GIT_INSTALL_ROOT=C:\tools\cmder\vendor\git-for-windows`` before I launch cmder.exe, then cmder can't find git.exe. The issue appears to be on line 186 of the init.bat:
``call :enhance_path "%GIT_INSTALL_ROOT%\usr\bin" append``
It should be
``call :enhance_path "%GIT_INSTALL_ROOT%\bin" append``

3. Trying to debug these issues, I had to go in and change init.bat to set ``debug-output=1``. I think it would be better to have the option to control this also by an env. var. the way it is in the current release. I.e. on lines 10 and 11, instead of:
``set verbose-output=0``
``set debug-output=0``
Having this would be useful:
``if not defined verbose-output ( set verbose-output=0 )``
``if not defined debug-output ( set debug-output=0 )``

**Comment by daxgames** [comment]
@i-s-o This versio of ini.bat supports command line arguments.  you can add a '/d' after init.bat in the task and it enables debug mode.  Look at the readme.md for init.bat command line arguments.

**Comment by DRSDavidSoft** [comment]
@i-s-o I agree with some of the changes you mentioned. We'd appreciate if you made a PR!
@daxgames Can you hold the merge before @i-s-o also confirms that the problems are fixed?

**Comment by i-s-o** [comment]
That didn't work for me because of the LF v. CRLF issue, so, I wanted to be able to see the debug output when I ran cmder.exe directly.

**Comment by i-s-o** [comment]
@DRSDavidSoft I will give it a try. This would be my first attempt at a PR (though, I've been following a couple of projects for a while now).

**Comment by daxgames** [comment]
@i-s-o what LF vs CRLF issue.

**Comment by i-s-o** [comment]
@daxgames init.bat's lines are terminated by LF (0x10). It seems like because of that, cmd.exe's batch processor can't find the subroutines. I am getting these types of errors:
``The system cannot find the batch label specified - enhance_path_recursive``
``The system cannot find the batch label specified - enhance_path``

**Comment by daxgames** [comment]
Thats really odd.  Is appveryor doing that?  I just downloaded the nightly and had the same issue.  If I run `unix2dos vendor\init.bat`  all works fine.

**Comment by Stanzilla** [comment]
@daxgames no, your branch currently has conflicts with master so it cannot be merged via a rebase, you have to rebase your changes onto current master.

**Comment by daxgames** [comment]
@Stanzilla I see no conflicts.

**Comment by daxgames** [comment]
@Stanzilla I am sorry and not ting to be argumentative but trying to understand.  Why would this branch show no conflicts on a merge and a squash and merge but shows conflicts on rebase and merge as you showed?

I personally have never used rebase and merge when incorporating a PR in github so why rebase and merge vs one of the other two options..

**Comment by i-s-o** [comment]
image

@daxgames, @DRSDavidSoft The way I read [this](https://stackoverflow.com/a/4095133), ``LF`` is not used as a token delimiter by the batch parser. Also, ``CALL`` scans the file on its own (i.e. while ``CR``s are still intact), so the parser thinks that ``LF`` is simply the last character of the batch label.

I'm going to test this again, though I have to figure out which appveyor build to use: I am guessing that I have to use the ones based on @daxgames's commits, because commits by others still have ``LF`` v. ``CRLF`` issue (at least until all the committers' ``.gittattributes`` gets updated as in https://github.com/cmderdev/cmder/pull/1665/commits/52db4449acf65b330ff4e0217e067c7500743242)

**Comment by Stanzilla** [comment]
@daxgames I think you just need to rebase onto master and then force push your branch, git is picky sometimes

**Comment by daxgames** [comment]
I screwed this up royally try to rebase - closing and trying to fix!

**Pull request closed without merging**